File: examples/python/webhook_handler.py
"""
Flask webhook handler with HMAC-SHA256 verification.

Usage:
    pip install flask
    PAPRE_WEBHOOK_SECRET=whsec_... python webhook_handler.py
"""
import hashlib
import hmac
import os
import json
from flask import Flask, request, jsonify

app = Flask(__name__)
WEBHOOK_SECRET = os.environ["PAPRE_WEBHOOK_SECRET"]


def verify_signature(payload: str, signature: str, secret: str) -> bool:
    """Verify the HMAC-SHA256 signature from the X-Papre-Signature header."""
    expected = hmac.new(secret.encode(), payload.encode(), hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, signature)


@app.route("/webhooks/papre", methods=["POST"])
def handle_webhook():
    payload = request.get_data(as_text=True)
    signature = request.headers.get("X-Papre-Signature", "")

    if not verify_signature(payload, signature, WEBHOOK_SECRET):
        return jsonify({"error": "Invalid signature"}), 401

    event = json.loads(payload)
    event_type = event["event_type"]

    if event_type == "agreement.signed":
        print(f"Agreement {event['agreement_id']} signed by {event['signer_name']}")
        print(f"Blockchain TX: {event.get('blockchain_tx')}")
    elif event_type == "agreement.viewed":
        print(f"Agreement {event['agreement_id']} viewed")
    elif event_type == "agreement.expired":
        print(f"Agreement {event['agreement_id']} expired")
    elif event_type == "agreement.declined":
        print(f"Agreement {event['agreement_id']} declined")
    elif event_type == "agreement.cancelled":
        print(f"Agreement {event['agreement_id']} cancelled")
    else:
        print(f"Unknown event: {event_type}")

    return jsonify({"received": True}), 200


if __name__ == "__main__":
    app.run(port=3000)

File: examples/node/webhook-handler.ts
/**
 * Express webhook handler with HMAC-SHA256 verification.
 *
 * Usage:
 *   PAPRE_WEBHOOK_SECRET=whsec_... npx tsx webhook-handler.ts
 */
import express from 'express';
import crypto from 'crypto';

const app = express();
const WEBHOOK_SECRET = process.env.PAPRE_WEBHOOK_SECRET!;

// Parse raw body for HMAC verification
app.post(
  '/webhooks/papre',
  express.raw({ type: 'application/json' }),
  (req, res) => {
    const signature = req.headers['x-papre-signature'] as string;
    const payload = req.body.toString();

    // Verify HMAC-SHA256 signature
    const expected = crypto
      .createHmac('sha256', WEBHOOK_SECRET)
      .update(payload)
      .digest('hex');

    if (!crypto.timingSafeEqual(Buffer.from(expected, 'hex'), Buffer.from(signature, 'hex'))) {
      console.error('Invalid webhook signature');
      return res.status(401).json({ error: 'Invalid signature' });
    }

    // Signature verified — process the event
    const event = JSON.parse(payload);

    switch (event.event_type) {
      case 'agreement.signed':
        console.log(`Agreement ${event.agreement_id} signed by ${event.signer_name}`);
        console.log(`Blockchain TX: ${event.blockchain_tx}`);
        // Update your database, trigger downstream actions, etc.
        break;

      case 'agreement.viewed':
        console.log(`Agreement ${event.agreement_id} viewed by signer`);
        break;

      case 'agreement.expired':
        console.log(`Agreement ${event.agreement_id} expired`);
        // Optionally recreate or notify admin
        break;

      case 'agreement.declined':
        console.log(`Agreement ${event.agreement_id} declined`);
        // Alert admin, flag participant
        break;

      case 'agreement.cancelled':
        console.log(`Agreement ${event.agreement_id} cancelled`);
        break;

      default:
        console.log(`Unknown event type: ${event.event_type}`);
    }

    // Always respond 200 quickly — Papre retries on non-2xx
    res.status(200).json({ received: true });
  },
);

app.listen(3000, () => {
  console.log('Webhook handler listening on port 3000');
});
